ffi/memorymanager.cpp

```diff
@@ -0,0 +1,334 @@
+//===---- memorymanager.cpp - Memory manager for MCJIT/RtDyld *- C++ -*----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the section-based memory manager used by the MCJIT
+// execution engine and RuntimeDyld
+//
+//===----------------------------------------------------------------------===//
+
+#include "memorymanager.h"
+#include "llvm/Support/MathExtras.h"
+#include "llvm/Support/Process.h"
+
+#define DEBUG_TYPE "llvmlite-memory-manager"
+
+namespace llvm {
+
+uint8_t *LlvmliteMemoryManager::allocateDataSection(uintptr_t Size,
+                                                    unsigned Alignment,
+                                                    unsigned SectionID,
+                                                    StringRef SectionName,
+                                                    bool IsReadOnly) {
+    if (IsReadOnly)
+        return allocateSection(LlvmliteMemoryManager::AllocationPurpose::ROData,
+                               Size, Alignment);
+    return allocateSection(LlvmliteMemoryManager::AllocationPurpose::RWData,
+                           Size, Alignment);
+}
+
+uint8_t *LlvmliteMemoryManager::allocateCodeSection(uintptr_t Size,
+                                                    unsigned Alignment,
+                                                    unsigned SectionID,
+                                                    StringRef SectionName) {
+    return allocateSection(LlvmliteMemoryManager::AllocationPurpose::Code, Size,
+                           Alignment);
+}
+
+uint8_t *LlvmliteMemoryManager::allocateSection(
+    LlvmliteMemoryManager::AllocationPurpose Purpose, uintptr_t Size,
+    unsigned Alignment) {
+    if (!Alignment)
+        Alignment = 16;
+
+    assert(!(Alignment & (Alignment - 1)) &&
+           "Alignment must be a power of two.");
+
+    uintptr_t RequiredSize =
+        Alignment * ((Size + Alignment - 1) / Alignment + 1);
```

Comment:
This computation is pushing a request for 16379 bytes with alignment 8 up to a size of 16392 bytes, which is larger than the 16384 bytes reserved, in the case where the assertion is being hit in the tests on M2.

Reply:
A couple of other things I just noticed:

- For `reserveAllocationSpace()`, the requested code size is 16380 bytes with an align of 4
- When the `allocateDataSection()` call is made for code, the request is for 16379 bytes with an align of 8

This discrepancy leads to the request being just slightly over what was reserved for code - I think the next step is to look into why the alignment and requested sizes differ.

Reply:
Output from the extra debugging prints I just pushed:

```
Code size / align: 0x3FFC / 4
ROData size / align: 0x0 / 1
RWData size / align: 0x2360 / 16
Reserving 0xC000 bytes
Code mem starts at 0x0000000132DDC000, size 0x4000
Rwdata mem starts at 0x0x0000000132DE0000, size 0x4000
Requested size / alignment: 0x3FFB / 8
Allocating 0x4008 bytes for CodeMem at Assertion failed: (false && "All memory must be pre-allocated"), function allocateSection, file memorymanager.cpp, line 109.
```

Reply:
[61ae2b0](https://github.com/numba/llvmlite/pull/1009/commits/61ae2b008f2b0b2ba19bd92120eee55d47e42612) appears to resolve this issue and allow the test to run to completion.